kpatch-build/create-diff-object.c

```diff
@@ -3676,6 +3681,206 @@ static void kpatch_create_callbacks_objname_rela(struct kpatch_elf *kelf, char *
+			insn_offset = sym->sym.st_value + PPC64_LOCAL_ENTRY_OFFSET(sym->sym.st_other);
+			insn = sym->sec->data->d_buf + insn_offset;
+
+			/* verify nops */
+			if (insn[0] != 0x00 || insn[1] != 0x00 || insn[2] != 0x00 || insn[3] != 0x60 ||
+			    insn[4] != 0x00 || insn[5] != 0x00 || insn[6] != 0x00 || insn[7] != 0x60) {
+				ERROR("%s: unexpected instruction in patch section of function\n", sym->name);
+			}
+
+			break;
+		case X86_64:
+			/*
+			 * Assume x86_64 is built with -fpatchable-function-entry=16,16, which means all 16 nops are
+			 * before the entry point of the function.  This is rather odd, but since call __fentry__ is
+			 * still used for the ftrace callsite, this option may only be currently used for alignment
+			 * purposes (and not ftrace as on other arches).
+			 *
+			 * Disassembly of section .text.cmdline_proc_show:
+			 *
+			 * 0000000000000000 <cmdline_proc_show-0x10>:
+			 *    0:   90                      nop                           << <<
+			 *    1:   90                      nop
+			 *    2:   90                      nop
+			 *    3:   90                      nop
+			 *    4:   90                      nop
+			 *    5:   90                      nop
+			 *    6:   90                      nop
+			 *    7:   90                      nop
+			 *    8:   90                      nop
+			 *    9:   90                      nop
+			 *    a:   90                      nop
+			 *    b:   90                      nop
+			 *    c:   90                      nop
+			 *    d:   90                      nop
+			 *    e:   90                      nop
+			 *    f:   90                      nop
+			 *
+			 * 0000000000000010 <cmdline_proc_show>:
+			 *   10:   f3 0f 1e fa             endbr64
+			 *   14:   e8 00 00 00 00          call   19 <cmdline_proc_show+0x9>
+			 *                         15: R_X86_64_PLT32      __fentry__-0x4
+			 *
+			 * Relocation section '.rela__patchable_function_entries' at offset 0x113f8 contains 1 entry:
+			 *     Offset             Info             Type               Symbol's Value  Symbol's Name + Addend
+			 * 0000000000000000  0000000700000001 R_X86_64_64            0000000000000000 .text.cmdline_proc_show + 0
+			 *                                                                                                      ^
+			 */
+			insn_offset = 0;
+			insn = sym->sec->data->d_buf + insn_offset;
+
+			/* verify nops */
+			for (int i=0; i<16; i++) {
+				if (insn[0] != 0x90) {
+					ERROR("%s: unexpected instruction in CONFIG_FUNCTION_PADDING_BYTES section of function\n", sym->name);
+				}
+				insn++;
+			}
+
+			/* optional endbr64 instruction */
+			if (insn[0] == 0xf3 && insn[1] == 0x0f && insn[2] == 0x1e && insn[3] == 0xfa)
+				insn+=4;
+
+			/*
+			 * We can't verify the ftrace call site as x86_64 adds nops (and
+			 * rela__patchable_function_entries) regardless of whether it is ftrace-able or not.
+			 * This arch still relies on __mcount_loc section, so we will check then.
+			 */
+			break;
```

Comment:
I believe all this x86-specific code can go away, as it doesn't actually use the section.  x86's use of `-fpatchable-function-entry=16,16` is a hack for prepending every function with nops.  Maybe the elf code can just not set kelf->has_pfe for x86.

Reply:
Fair enough.  x86_64's (ab)use of these sections is rather strange, so if their omission given a .config / compiler that would otherwise create them in a kernel object file is safe, then I can cut it out.